mava/evaluator.py

```diff
@@ -12,329 +12,140 @@
+                not evenly divide `eval_episodes`
     """
-    # Check if win rate is required for evaluation.
-    log_win_rate = config.env.log_win_rate
-    # Vmap it over number of agents and create evaluator_fn.
-    if use_recurrent_net:
-        assert scanned_rnn is not None
-        evaluator = get_rnn_evaluator_fn(
-            eval_env,
-            network_apply_fn,  # type: ignore
-            config,
-            scanned_rnn,
-            log_win_rate,
-        )
-        absolute_metric_evaluator = get_rnn_evaluator_fn(
-            eval_env,
-            network_apply_fn,  # type: ignore
-            config,
-            scanned_rnn,
-            log_win_rate,
-            10,
-        )
-    else:
-        evaluator = get_ff_evaluator_fn(
-            eval_env, network_apply_fn, config, log_win_rate  # type: ignore
-        )
-        absolute_metric_evaluator = get_ff_evaluator_fn(
-            eval_env, network_apply_fn, config, log_win_rate, 10  # type: ignore
-        )
-
-    evaluator = jax.pmap(evaluator, axis_name="device")
-    absolute_metric_evaluator = jax.pmap(absolute_metric_evaluator, axis_name="device")
-
-    return evaluator, absolute_metric_evaluator
+
+    def eval_fn(params: FrozenDict, key: PRNGKey, init_act_state: ActorState) -> Dict[str, Array]:
+        def _env_step(eval_state: _EvalEnvStepState, _: Any) -> Tuple[_EvalEnvStepState, TimeStep]:
+            """Performs a single environment step"""
+            env_state, ts, key, actor_state = eval_state
+
+            key, act_key = jax.random.split(key)
+            action, actor_state = act_fn(params, ts, act_key, **actor_state)
+            env_state, ts = jax.vmap(env.step)(env_state, action)
+
+            return (env_state, ts, key, actor_state), ts
+
+        def _episode(key: PRNGKey, _: Any) -> Tuple[PRNGKey, Metrics]:
+            """Simulates `config.arch.num_envs` episodes."""
+            key, reset_key = jax.random.split(key)
+            reset_keys = jax.random.split(reset_key, config.arch.num_envs)
+            env_state, ts = jax.vmap(env.reset)(reset_keys)
+
+            step_state = env_state, ts, key, init_act_state
+            _, timesteps = jax.lax.scan(_env_step, step_state, jnp.arange(env.time_limit))
+
+            metrics = timesteps.extras["episode_metrics"]
+            if config.env.log_win_rate:
+                metrics["won_episode"] = timesteps.extras["won_episode"]
+
+            # find the first instance of done to get the metrics at that timestep we don't
+            # care about subsequent steps because we only the results from the first episode
+            done_idx = jnp.argmax(timesteps.last(), axis=0)
+            metrics = jax.tree_map(lambda m: m[done_idx, jnp.arange(config.arch.num_envs)], metrics)
+            del metrics["is_terminal_step"]  # uneeded for logging
+
+            return key, metrics
+
+        # todo: do we want to divide this by n devices?
```

Comment:
About this todo, I think yes we need to devide them over devices and then over num_envs

Reply:
I think we need to make a warning in case math.ceil(eval_episodes/num_envs) is not equal eval_episodes/num_envs

Reply:
Ye I agree, although we weren't doing this before the refactor.

Agreed on the warning, I like the idea

Reply:
We were doing this before: `eval_batch = (config.arch.num_eval_episodes // n_devices) * eval_multiplier`

Reply:
ah ok I missed that, then I must definitely divide by the devices